src/preamble.js

```diff
@@ -550,35 +559,42 @@ function UTF8ArrayToString(u8Array, idx) {
       str += String.fromCharCode(0xD800 | (ch >> 10), 0xDC00 | (ch & 0x3FF));
     }
   }
+  return str;
 }
 {{{ maybeExport('UTF8ArrayToString') }}}
 
-// Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the emscripten HEAP, returns
+// Given a pointer 'ptr' to an UTF8-encoded string (optionally null terminated) in the emscripten HEAP, returns
 // a copy of that string as a Javascript String object.
+// len parameter (optional): bytewise length in case string is not null terminated
 
-function UTF8ToString(ptr) {
-  return UTF8ArrayToString({{{ heapAndOffset('HEAPU8', 'ptr') }}});
+function UTF8ToString(ptr, len) {
+  return UTF8ArrayToString({{{ heapAndOffset('HEAPU8', 'ptr') }}}, len);
 }
 {{{ maybeExport('UTF8ToString') }}}
 
 // Copies the given Javascript String object 'str' to the given byte array at address 'outIdx',
-// encoded in UTF8 form and null-terminated. The copy will require at most str.length*4+1 bytes of space in the HEAP.
+// encoded in UTF8 form and optionally null-terminated. The copy will require at most str.length*4+1 bytes of space in the HEAP.
 // Use the function lengthBytesUTF8() to compute the exact number of bytes (excluding null terminator) that this function will write.
 // Parameters:
 //   str: the Javascript string to copy.
 //   outU8Array: the array to copy to. Each index in this array is assumed to be one 8-byte element.
 //   outIdx: The starting offset in the array to begin the copying.
-//   maxBytesToWrite: The maximum number of bytes this function can write to the array. This count should include the null 
-//                    terminator, i.e. if maxBytesToWrite=1, only the null terminator will be written and nothing else.
+//   maxBytesToWrite: The maximum number of bytes this function can write to the array. 
+//                    When null terminator is used (see skipNullTerminator): This count should include the null terminator, 
+//                    i.e. if maxBytesToWrite=1, only the null terminator will be written and nothing else.
 //                    maxBytesToWrite=0 does not write any bytes to the output, not even the null terminator.
-// Returns the number of bytes written, EXCLUDING the null terminator.
+//   skipNullTerminator: request null terminator to be omitted; optional, defaults to false
+// Returns the number of bytes written, EXCLUDING the optional null terminator.
 
-function stringToUTF8Array(str, outU8Array, outIdx, maxBytesToWrite) {
+function stringToUTF8Array(str, outU8Array, outIdx, maxBytesToWrite, skipNullTerminator) {
```

Comment:
This is another case where you might want a new function instead of dynamically checking the skipNullTerminator argument.